compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileTensor.cpp

```diff
@@ -0,0 +1,183 @@
+                                      SmallVectorImpl<int64_t> &workgroupSize,
+                                      bool distributeToWarp) {
+  std::array<int64_t, 3> elementPerWorkgroup = {
+      distributeToWarp ? workgroupSize[0] / kWarpSize : workgroupSize[0],
+      workgroupSize[1], workgroupSize[2]};
+  SmallVector<Operation *> computeOps;
+  SmallVector<LoopTilingAndDistributionInfo> tiledLoops;
+  if (failed(getComputeOps(funcOp, computeOps, tiledLoops))) {
+    return funcOp.emitOpError("failed to get compute ops");
+  }
+
+  for (Operation *op : computeOps) {
+    auto tilingOp = dyn_cast<TilingInterface>(op);
+    if (!tilingOp) continue;
+    size_t numLoops = 0;
+    for (auto type : tilingOp.getLoopIteratorTypes()) {
+      if (type == getParallelIteratorTypeName()) numLoops++;
+    }
+    IRRewriter rewriter(op->getContext());
+    rewriter.setInsertionPoint(op);
+    auto interfaceOp = cast<PartitionableLoopsInterface>(*op);
+    auto partitionedLoops =
+        interfaceOp.getPartitionableLoops(kNumMaxParallelDims);
+    SmallVector<OpFoldResult> numThreads(numLoops, rewriter.getIndexAttr(0));
+    int64_t id = 0;
+    int64_t threadId = 0;
+    SmallVector<int64_t> blockTileSize = getTileSizes(op, 0);
+    SmallVector<int64_t> idDims;
+    for (unsigned loop : llvm::reverse(partitionedLoops)) {
+      if (loop >= blockTileSize.size()) continue;
+      int64_t num = elementPerWorkgroup[id++];
+      if (num > 1) {
+        numThreads[loop] = rewriter.getIndexAttr(num);
+        idDims.push_back(threadId++);
+      }
+    }
+    std::reverse(idDims.begin(), idDims.end());
+    for (int64_t i = threadId; i < 3; i++) idDims.push_back(i);
+
+    auto tilingResult =
+        linalg::tileToForeachThreadOp(rewriter, tilingOp, numThreads, idDims);
+    rewriter.replaceOp(op, tilingResult->tileOp->getResults());
+  }
+  return success();
+}
+
+namespace {
+struct LLVMGPUTileTensorPass
+    : public LLVMGPUTileTensorBase<LLVMGPUTileTensorPass> {
+ private:
+  // Distribute the workloads to warp if true otherwise distribute to threads.
+  bool distributeToWarp = false;
+
+ public:
+  LLVMGPUTileTensorPass(bool distributeToWarp)
+      : distributeToWarp(distributeToWarp) {}
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AffineDialect, gpu::GPUDialect>();
+  }
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+    if (!isEntryPoint(funcOp)) return;
+
+    auto workgroupSize = llvm::to_vector<4>(llvm::map_range(
+        getEntryPoint(funcOp)->getWorkgroupSize().getValue(),
+        [&](Attribute attr) { return attr.cast<IntegerAttr>().getInt(); }));
+    if (failed(tileParallelDims(funcOp, workgroupSize, distributeToWarp))) {
+      return signalPassFailure();
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "--- After second level of tiling";
+      funcOp.dump();
+    });
+
+    if (failed(tileReduction(funcOp))) {
```

Comment:
This one is very iffy for me.
Atm `scf.foreach_thread` does not have reduction semantics, this is still WIP.

Reply:
Actually, this is not even tested. 
Can we drop the reduction part until there is proper support for it?
This should still compose with other ways of performing reduction that can be called after this.

Reply:
The reduction tiling is done using `scf.for`, it is needed to be able to vectorize reductions without creating large vectors. I added a test for it. Should be fine I believe.